v-next/hardhat/src/internal/builtin-plugins/solidity-test/reporter.ts

```diff
@@ -119,33 +120,48 @@ export async function* testReporter(
     for (const [index, failure] of failures.entries()) {
       yield `\n${chalk.bold(chalk.red(`Failure (${index + 1})`))}: ${failure.name}\n`;
 
-      if (
-        failure.reason !== undefined &&
-        failure.reason !== null &&
-        failure.reason !== ""
-      ) {
-        yield `Reason: ${chalk.grey(failure.reason)}\n`;
+      const stackTrace = failure.stackTrace();
+
+      let reason: string | undefined;
+      if (stackTrace?.kind === "StackTrace") {
+        reason = getMessageFromLastStackTraceEntry(
+          stackTrace.entries[stackTrace.entries.length - 1],
+        );
+      }
+      if (reason === undefined || reason === "") {
+        reason = failure.reason ?? "Unknown error";
       }
 
-      const stackTrace = failure.stackTrace();
-      // TODO handle `UnexpectedError` and `UnsafeToReplay` variants
-      if (
-        stackTrace !== undefined &&
-        stackTrace !== null &&
-        stackTrace.kind === "StackTrace" &&
-        stackTrace.entries.length > 0
-      ) {
-        const stackTraceStack: string[] = [];
-        for (const entry of stackTrace.entries.reverse()) {
-          const callsite = encodeStackTraceEntry(entry);
-          if (callsite !== undefined) {
-            stackTraceStack.push(`  at ${callsite.toString()}`);
+      yield `Reason: ${chalk.grey(reason)}\n`;
+
+      switch (stackTrace?.kind) {
+        case "StackTrace":
+          const stackTraceStack: string[] = [];
+          for (const entry of stackTrace.entries.reverse()) {
+            const callsite = encodeStackTraceEntry(entry);
+            if (callsite !== undefined) {
+              stackTraceStack.push(`  at ${callsite.toString()}`);
+            }
           }
-        }
 
-        if (stackTraceStack.length > 0) {
-          yield `${chalk.grey(stackTraceStack.join("\n"))}\n`;
-        }
+          if (stackTraceStack.length > 0) {
+            yield `${chalk.grey(stackTraceStack.join("\n"))}\n`;
+          }
+          break;
+        case "UnexpectedError":
+          yield `Stack Trace Warning: ${chalk.grey(stackTrace.errorMessage)}\n`;
+          break;
+        case "UnsafeToReplay":
+          if (stackTrace.globalForkLatest === true) {
+            yield `Stack Trace Warning: ${chalk.grey("The test is not safe to replay because a fork url without a fork block number was provided.")}\n`;
+          }
+          if (stackTrace.impureCheatcodes.length > 0) {
+            yield `Stack Trace Warning: ${chalk.grey(`The test is not safe to replay because it uses impure cheatcodes: ${stackTrace.impureCheatcodes.join(", ")}`)}\n`;
+          }
+          break;
+        case "HeuristicFailed":
```

Comment:
What does this mean, @agostbiro? Does it make sense to just ignore it? How is it different than `UnexpectedError`?

Reply:
We get `HeuristicFailed` when the process of generating stack trace goes through, but the stack trace is empty i.e. we couldn't find any useful information. We don't get any additional information on why that might have been though. We could inform the user about it but I'm not sure if there's anything actionable on their part we could display.

`UnexpectedError`, on the other hand, happens when the stack trace generation throws.